File: ios-native/FlynnAI/Core/KeyboardBridge.swift
import Foundation

/// Bridges the main app and the keyboard extension. The sandboxed keyboard can't
/// run the Supabase SDK or refresh short-lived tokens, so the app mints a
/// long-lived keyboard JWT (`/api/keyboard/provision-token`) and stashes it, plus
/// the API base URL and business name, in the shared App Group / keychain for the
/// keyboard to use. Call on launch, on foreground, and after sign-in.
enum KeyboardBridge {

    private struct ProvisionTokenResponse: Decodable {
        let token: String
        let expiresAt: String?
    }

    /// Refresh everything the keyboard needs. Best-effort: never throws.
    static func sync(businessName: String? = nil) async {
        // Always publish the API base URL so the keyboard can reach the backend.
        SharedStore.apiBaseURL = FlynnEnv.flynnAPIBaseURL.absoluteString
        if let businessName, !businessName.isEmpty {
            SharedStore.businessName = businessName
        }

        do {
            let session = try await FlynnSupabase.client.auth.session
            var req = URLRequest(
                url: FlynnEnv.flynnAPIBaseURL.appendingPathComponent("api/keyboard/provision-token")
            )
            req.httpMethod = "POST"
            req.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = Data("{}".utf8)

            let (data, response) = try await URLSession.shared.data(for: req)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
                return
            }
            let decoded = try JSONDecoder().decode(ProvisionTokenResponse.self, from: data)
            SharedSecureStore.setKeyboardToken(decoded.token)
        } catch {
            // No session yet, or network error — the keyboard shows its
            // "open Flynn to finish setup" state until this succeeds.
        }
    }

    /// Clear the keyboard's access on sign-out.
    static func clear() {
        SharedSecureStore.delete(account: FlynnShared.keyboardTokenAccount)
    }
}

File: android-native/app/src/main/java/com/flynnai/app/data/api/KeyboardBridge.kt
package com.flynnai.app.data.api

import android.content.Context
import com.flynnai.app.FlynnApp
import com.flynnai.app.core.Environment
import com.flynnai.app.keyboard.KeyboardTokenStore
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.status.SessionStatus

// Mirrors iOS KeyboardBridge.sync() — call on app foreground + after sign-in.
object KeyboardBridge {

    suspend fun sync(context: Context, businessName: String? = null) {
        val client = FlynnApp.instance.supabase

        // Always write the API base URL so the IME can find the backend
        KeyboardTokenStore.setApiBase(context, Environment.flynnApiBaseUrl)
        if (!businessName.isNullOrBlank()) {
            KeyboardTokenStore.setBusinessName(context, businessName)
        }

        try {
            val status = client.auth.sessionStatus.value
            if (status !is SessionStatus.Authenticated) return
            val accessToken = client.auth.currentAccessTokenOrNull() ?: return
            val token = FlynnApi.provisionKeyboardToken(accessToken)
            KeyboardTokenStore.setToken(context, token)
        } catch (_: Exception) {
            // Best-effort — IME shows "open Flynn to finish setup" until this succeeds
        }
    }

    fun clear(context: Context) = KeyboardTokenStore.clear(context)
}
